### PR TITLE
Putty-Style-Mousebutton-Swap

### PR DESCRIPTION
**Pull request by macrl2000**
Hi, i tried this years ago for some ancient branch which made no sense to integrate, but recently hat to redo it for the latest version. i have zero clue if this will break things for you, if it fits in your coding style, or if you like it at all, but i am using this a lot, so i am making this pull request, feel free to deny, modify or use it :-)

this pull request adds the configurable feature that you can swap middle and right mouse button application side.

### CODE REVIEW

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -658,7 +658,7 @@ void MainWindow::actAbout_triggered()
 {
      QMessageBox::about(this, tr("About"),
                      QStringLiteral("<center><b><big>QTerminal %1</big></b></center><br>").arg(qApp->applicationVersion())
-                     + tr("A lightweight and powerful multiplatform terminal emulator")
+                     + tr("A lightweight and powerful multiplatform terminal emulator (mouse swap build)")
```
Comment (yan12125) [inline src/mainwindow.cpp:661]: This feature will become part of the official version, so a cooment is no needed.

File: src/properties.cpp [hunk 2]
```diff
@@ -164,6 +164,8 @@ void Properties::loadSettings()
 
     windowMaximized = m_settings->value(QLatin1String("LastWindowMaximized"), false).toBool();
 
+    SwapMouseButtons2and3 = m_settings->value(QLatin1String("SwapMouseButtons2and3"), false).toBool();
```
Comment (yan12125) [inline src/properties.cpp:167]: Coding style: the first character of a variable name is generally not capitalized.

File: src/termwidget.cpp [hunk 3]
```diff
@@ -76,8 +76,17 @@ TermWidgetImpl::TermWidgetImpl(TerminalConfig &cfg, QWidget * parent)
     disableBracketedPasteMode(Properties::Instance()->m_disableBracketedPasteMode);
 
     setContextMenuPolicy(Qt::CustomContextMenu);
-    connect(this, &QWidget::customContextMenuRequested,
-            this, &TermWidgetImpl::customContextMenuCall);
+
+    if(Properties::Instance()->SwapMouseButtons2and3 == true)
```
Comment (yan12125) [inline src/termwidget.cpp:80]: `== true` is not needed

File: src/termwidget.cpp [hunk 4]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
     }
 }
 
+bool TermWidget::eventFilter(QObject * /*obj*/, QEvent * ev)
+{
+    if (ev->type() == QEvent::MouseButtonPress)
+    {
+        QMouseEvent *mev = (QMouseEvent *)ev;
```
Comment (yan12125) [inline src/termwidget.cpp:254]: C-style casts are less than ideal. Could you change this to xxx_cast?

File: src/termwidget.cpp [hunk 5]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
...
+        if ( mev->button() == Qt::MiddleButton )
+        {
+            impl()->pasteSelection();
+            if(Properties::Instance()->SwapMouseButtons2and3 == true)
+            {
+                    const QPoint &point = (const QPoint &)(mev->pos());
```
Comment (yan12125) [inline src/termwidget.cpp:260]: Ditto for C-style cast.

File: src/termwidget.cpp [hunk 6]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
...
 
+bool TermWidget::eventFilter(QObject * /*obj*/, QEvent * ev)
+{
+    if (ev->type() == QEvent::MouseButtonPress)
+    {
+        QMouseEvent *mev = (QMouseEvent *)ev;
+        if ( mev->button() == Qt::MiddleButton )
+        {
+            impl()->pasteSelection();
+            if(Properties::Instance()->SwapMouseButtons2and3 == true)
```
Comment (yan12125) [inline src/termwidget.cpp:258]: Ditto for `== true`

File: src/properties.h [hunk 7]
```diff
@@ -124,6 +124,7 @@ class Properties
         bool trimPastedTrailingNewlines;
 
         bool windowMaximized;
+	bool SwapMouseButtons2and3;
```
Comment (yan12125) [inline src/properties.h:127]: Ditto for variable naming.

**Review by yan12125 — COMMENTED** [review]
Thank you for this nice feature! At first glance it looks good. Just some minor stuffs.

**Comment by macrl2000** [comment]
i pushed the code changes you requested, hope this is all correct now

**Comment by yan12125** [comment]
There are some compile errors as shown in https://github.com/lxqt/qterminal/actions/runs/6310403117/job/17137137707. Could you check?

**Comment by macrl2000** [comment]
i had one letter missing in properties.h sorry, i recompiled - should work now

**Comment by macrl2000** [comment]
sorry i need to check what went wrong - i cannot use git correctly :-(

**Comment by macrl2000** [comment]
maybe now ?

**Comment by macrl2000** [comment]
anything left to do for me ? or can you use it now ?

File: src/termwidget.cpp [hunk 8]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
...
 }
 
+bool TermWidget::eventFilter(QObject * /*obj*/, QEvent * ev)
+{
+    if (ev->type() == QEvent::MouseButtonPress)
+    {
+        QMouseEvent *mev = static_cast<QMouseEvent*>(ev);
+        if ( mev->button() == Qt::MiddleButton )
+        {
+            impl()->pasteSelection();
```
Comment (yan12125) [inline src/termwidget.cpp:257]: This line seems unneeded?

File: src/termwidget.cpp [hunk 9]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
...
+            if(Properties::Instance()->swapMouseButtons2and3)
+            {
+                    const QPoint& point = static_cast<const QPoint&>(mev->pos());
+                    impl()->customContextMenuCall(point);
+            }
+            else  impl()->pasteSelection();
```
Comment (yan12125) [inline src/termwidget.cpp:263]: Please also add curly brackets for one-line block.

File: src/termwidget.cpp [hunk 10]
```diff
@@ -254,6 +286,15 @@ TermWidget::TermWidget(TerminalConfig &cfg, QWidget * parent)
     setLayout(m_layout);
 
     m_layout->addWidget(m_term);
+    const auto objs = m_term->children();
+
+    for (QObject *o : objs)
+    {
+        // Find TerminalDisplay
+        if (!o->isWidgetType() || qobject_cast<QWidget*>(o)->isHidden())
+            continue;
```
Comment (yan12125) [inline src/termwidget.cpp:295]: Ditto for curly brackets. Indeed the original code doesn't include brackets. Let's improve the code when you're at it.

File: src/termwidget.cpp [hunk 11]
```diff
@@ -238,6 +247,29 @@ void TermWidgetImpl::bell() {
...
+{
+    if (ev->type() == QEvent::MouseButtonPress)
+    {
+        QMouseEvent *mev = static_cast<QMouseEvent*>(ev);
+        if ( mev->button() == Qt::MiddleButton )
+        {
+            impl()->pasteSelection();
+            if(Properties::Instance()->swapMouseButtons2and3)
+            {
+                    const QPoint& point = static_cast<const QPoint&>(mev->pos());
```
Comment (yan12125) [inline src/termwidget.cpp:260]: Is a cast necesary here? These two lines can also be combined as `impl()->customContextMenuCall(mev->pos());`

File: src/properties.h [hunk 12]
```diff
@@ -124,6 +124,7 @@ class Properties
         bool trimPastedTrailingNewlines;
 
         bool windowMaximized;
+	bool swapMouseButtons2and3;
```
Comment (yan12125) [inline src/properties.h:127]: Don't use tabs when a file uses spaces.

File: src/termwidget.cpp [hunk 13]
```diff
@@ -76,8 +76,17 @@ TermWidgetImpl::TermWidgetImpl(TerminalConfig &cfg, QWidget * parent)
     disableBracketedPasteMode(Properties::Instance()->m_disableBracketedPasteMode);
 
     setContextMenuPolicy(Qt::CustomContextMenu);
-    connect(this, &QWidget::customContextMenuRequested,
-            this, &TermWidgetImpl::customContextMenuCall);
+
+    if(Properties::Instance()->swapMouseButtons2and3)
+    {
+	connect(this, &QWidget::customContextMenuRequested,
+		this, &TermWidgetImpl::pasteSelection);
```
Comment (yan12125) [inline src/termwidget.cpp:83]: Ditto for tabs.

File: src/termwidget.cpp [hunk 14]
```diff
@@ -76,8 +76,17 @@ TermWidgetImpl::TermWidgetImpl(TerminalConfig &cfg, QWidget * parent)
...
+    }
+    else
+    {
+	connect(this, &QWidget::customContextMenuRequested,
+		this, &TermWidgetImpl::customContextMenuCall);
```
Comment (yan12125) [inline src/termwidget.cpp:88]: Ditto for tabs.

File: src/termwidget.h [hunk 15]
```diff
@@ -87,6 +87,8 @@ class TermWidget : public QWidget, public DBusAddressable
         void closeTerminal();
         #endif
 
+	bool eventFilter(QObject * obj, QEvent * evt) override;
```
Comment (yan12125) [inline src/termwidget.h:90]: Ditto for tabs.

File: src/termwidget.h [hunk 16]
```diff
@@ -50,9 +50,9 @@ class TermWidgetImpl : public QTermWidget
         void zoomIn();
         void zoomOut();
         void zoomReset();
+	void customContextMenuCall(const QPoint & pos);
```
Comment (yan12125) [inline src/termwidget.h:53]: Ditto for tabs.

**Review by yan12125 — COMMENTED** [review]
Please also add a GUI option, probably in the "Behavior" tag of the preferences dialog.

**Comment by macrl2000** [comment]
> Please all add a GUI option, probably in the "Behavior" tag of the preferences dialog.

for this my qt coding skills are not sufficient, you can maybe add yourself or leave it as config-file only

**Comment by yan12125** [comment]
Here we go!

> for this my qt coding skills are not sufficient, you can maybe add yourself or leave it as config-file only

No problem, a GUI option can be added via another pull request.

**Pull request merged**